website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+        conn.execute('ROLLBACK TO sql_check;')
+        print("---SQL QUERY VALIDATION: SUCCESS---")
+    except Exception as e:
+        # Roll back in case of error
+        conn.execute('ROLLBACK TO sql_check;')
+        print("---SQL QUERY VALIDATION: FAILED---")
+        print(f"Error: {e}")
+        messages += [("user", f"Your SQL query failed to execute: {e}")]
+        error = "yes"
+
+    state["error"] = error
+
+    return state
+```
+#### **7. Run Query**
+
+The `run_query` node executes the validated SQL query and retrieves the results.
+
+**Purpose:** Execute the SQL query and provide the output.
+
+**Process:**
+
+- Connects to the database to run the query.
+
+- Fetches results and updates the state with the retrieved data.
+
+**Key Considerations:**
+
+- Implements robust error handling for execution issues.
+
+- Ensures results are in a usable format for further analysis or reporting.
+
+**Examples:**
+
+- Input SQL: _"SELECT COUNT(*) FROM Customers WHERE City = 'New York';"_
+
+- Query Result: _"(0,)"_
+
+- Input SQL: _"SELECT SUM(Products.Price * OrderDetails.Quantity) AS TotalSales FROM OrderDetails LEFT JOIN Products ON OrderDetails.ProductID = Products.ProductID;"*
+
+- Query Result: _"(6925.0,)"_
+
+**Code:**
+
+```python
+def run_query(state: GraphState):
+    print("---RUNNING SQL QUERY---")
+    sql_solution = state["generation"]
+    sql_code = sql_solution.sql_code.strip()
+    results = None
+    no_records_found = False  # Flag to indicate no records found
+
+    try:
+        # Execute the SQL query
+        cursor.execute(sql_code)
+        # For SELECT queries, fetch and store results
+        if sql_code.upper().startswith("SELECT"):
+            results = cursor.fetchall()
+            if not results:
+                no_records_found = True
+                print("---SQL QUERY EXECUTION: SUCCESS---")
+                print("No records found.")
+            else:
+                print("---SQL QUERY EXECUTION: SUCCESS---")
+        else:
+            # For non-SELECT queries, commit the changes
+            conn.commit()
+            print("---SQL QUERY EXECUTION: SUCCESS---")
+    except Exception as e:
+        print("---SQL QUERY EXECUTION: FAILED---")
+        print(f"Error: {e}")
+
+    # Store the results and no_records_found flag in the state
+    state["results"] = results
+    state["no_records_found"] = no_records_found
+    return state
+```
+
+The schema above provides a representation of the LangGraph Nodes and Edges:
+
+![graph_representation](graph.png)
+
+# Logging the Model in MLFlow
+
+Now that we have built a Multi-Lingual Query Engine using LangGraph, we are ready to log the model using MLflow’s [ Model from Code](https://mlflow.org/blog/models_from_code). This approach, where we log the code that represents the model, contrasts with object-based logging, where a model object is created, serialized, and logged as a pickle or JSON object.
```

Comment:
Motivate this step (logging the model) a little more? e.g. can mention versioning, sharing, packaging for deployment, etc.

Reply:
I've expanded the section to better motivate the model logging step, emphasizing how MLflow enables tracking, versioning, and packaging for deployment across different serving infrastructures.